In [1]:
import pandas as pd
import numpy as np

In [2]:
dados= pd.read_excel("/Users/camil/Jupyter Projects/dados/Kaggle_Sirio_Libanes_ICU_Prediction.xlsx")

In [3]:
dados

PATIENT_VISIT_IDENTIFIER  AGE_ABOVE65 AGE_PERCENTIL  GENDER  \
0                            0            1          60th       0   
1                            0            1          60th       0   
2                            0            1          60th       0   
3                            0            1          60th       0   
4                            0            1          60th       0   
...                        ...          ...           ...     ...   
1920                       384            0          50th       1   
1921                       384            0          50th       1   
1922                       384            0          50th       1   
1923                       384            0          50th       1   
1924                       384            0          50th       1   

      DISEASE GROUPING 1  DISEASE GROUPING 2  DISEASE GROUPING 3  \
0                    0.0                 0.0                 0.0   
1                    0.0                 0.0                 0.0   
2                    0.0                 0.0                 0.0   
3                    0.0                 0.0                 0.0   
4                    0.0                 0.0                 0.0   
...                  ...                 ...                 ...   
1920                 0.0                 0.0                 0.0   
1921                 0.0                 0.0                 0.0   
1922                 0.0                 0.0                 0.0   
1923                 0.0                 0.0                 0.0   
1924                 0.0                 0.0                 1.0   

      DISEASE GROUPING 4  DISEASE GROUPING 5  DISEASE GROUPING 6  ...  \
0                    0.0                 1.0                 1.0  ...   
1                    0.0                 1.0                 1.0  ...   
2                    0.0                 1.0                 1.0  ...   
3                    0.0                 1.0                 1.0  ...   
4                    0.0                 1.0                 1.0  ...   
...                  ...                 ...                 ...  ...   
1920                 0.0                 0.0                 0.0  ...   
1921                 0.0                 0.0                 0.0  ...   
1922                 0.0                 0.0                 0.0  ...   
1923                 0.0                 0.0                 0.0  ...   
1924                 0.0                 0.0                 0.0  ...   

      TEMPERATURE_DIFF  OXYGEN_SATURATION_DIFF  \
0            -1.000000               -1.000000   
1            -1.000000               -1.000000   
2                  NaN                     NaN   
3            -1.000000               -1.000000   
4            -0.238095               -0.818182   
...                ...                     ...   
1920         -1.000000               -1.000000   
1921         -1.000000               -1.000000   
1922         -1.000000               -1.000000   
1923         -1.000000               -1.000000   
1924         -0.547619               -0.838384   

      BLOODPRESSURE_DIASTOLIC_DIFF_REL  BLOODPRESSURE_SISTOLIC_DIFF_REL  \
0                            -1.000000                        -1.000000   
1                            -1.000000                        -1.000000   
2                                  NaN                              NaN   
3                                  NaN                              NaN   
4                            -0.389967                         0.407558   
...                                ...                              ...   
1920                         -1.000000                        -1.000000   
1921                         -1.000000                        -1.000000   
1922                         -1.000000                        -1.000000   
1923                         -1.000000                        -1.000000   
1924                         -0.701863                   

Já sabemos que este dataset tem diferentes features, sendo que elas contém dados sobre: informações demográficas, doenças pré-existentes, exames de sangue e sinais vitais. Além disso temos as colunas 'WINDOW' que indica o tempo desde a entrada do paciente no hospital e'ICU' que é a variável alvo. Vamos fazer uma avaliação dessas variáveis:

In [19]:
dados.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1925 entries, 0 to 1924
Columns: 231 entries, PATIENT_VISIT_IDENTIFIER to ICU
dtypes: float64(225), int64(4), object(2)
memory usage: 3.4+ MB


São 231 colunas, sendo duas do tipo object e que portanto não vão permitir que o modelo rode.

In [8]:
dados['PATIENT_VISIT_IDENTIFIER'].nunique()

385

Tem 385 pacientes diferentes no dataset

In [15]:
dados.columns

Index(['PATIENT_VISIT_IDENTIFIER', 'AGE_ABOVE65', 'AGE_PERCENTIL', 'GENDER',
       'DISEASE GROUPING 1', 'DISEASE GROUPING 2', 'DISEASE GROUPING 3',
       'DISEASE GROUPING 4', 'DISEASE GROUPING 5', 'DISEASE GROUPING 6',
       ...
       'TEMPERATURE_DIFF', 'OXYGEN_SATURATION_DIFF',
       'BLOODPRESSURE_DIASTOLIC_DIFF_REL', 'BLOODPRESSURE_SISTOLIC_DIFF_REL',
       'HEART_RATE_DIFF_REL', 'RESPIRATORY_RATE_DIFF_REL',
       'TEMPERATURE_DIFF_REL', 'OXYGEN_SATURATION_DIFF_REL', 'WINDOW', 'ICU'],
      dtype='object', length=231)

In [20]:
colunas =dados.describe().columns #não pega as colunas type object


In [21]:
for column in dados.columns:
    if column not in dados[colunas].columns:
        print(column)

AGE_PERCENTIL
WINDOW


As colunas do tipo object são  as colunas AGE_PERCENTIL E WINDOW. É necessário alterar o tipo de dado dessas colunas se quiser utilizá-las no modelo. A coluna WINDOW é a que fala o tempo em horas entre a admissão do paciente e o momento dos resultados. Como queremos que o modelo preveja com rapidez a necessidade ou não de UTI, apenas os dados da primeira janela de tempo serão utilizados. Além disso, se nessa janela o paciente já foi para UTI, os dados desse paciente serão descartados completamente, pois os exames podem ter sido feito após o paciente ter sido encaminhado à UTI, conforme informado pela equipe do Sírio Libanês. 

Um outro problema que é necessário resolver, é a questão dos dados faltantes. Nem todos os resultados estão preenchidos para todos os pacientes e janelas. Isso acontece porque os exames podem não ser repetidos de uma janela para outra. Sinais vitais podem ser medidos diversas vezes ao dia, mas exames de sangue não costumam ser coletados a todo o momento. O Hospital considera razoável considerar que se não há uma nova medida de uma variável, é muito provável que os valores seriam similares aos medidos nas janelas de tempo vizinhas.

Para resolver esses problemas será implementada uma função que recebe os dados de sinais vitais e exames e agrupa por paciente para fazer o preenchimento dos dados faltantes utilizando os dados das janelas anteriores e posteriores contanto que o paciente não tenha sido admitido na UTI nessas janelas.

In [22]:
def preenche_tabela(dados):
    def preenche_na(rows):
        rows.loc[rows["ICU"] != 1] = rows.loc[rows["ICU"] != 1].fillna(method='bfill').fillna(method='ffill')#resolução da Alura do problema do bfill
        return rows
    
    features_continuas_colunas = dados.iloc[:, 13:-2].columns
    features_continuas = dados.groupby("PATIENT_VISIT_IDENTIFIER", as_index=False)[list(features_continuas_colunas) + ["ICU"]].apply(preenche_na)
    features_continuas.drop("ICU", axis=1, inplace=True)
    
    features_categoricas = dados.iloc[:, :13]
    saida = dados.iloc[:, -2:]
    
    dados_finais = pd.concat([features_categoricas, features_continuas, saida], ignore_index=True, axis=1)
    dados_finais.columns = dados.columns
    
    return dados_finais

In [34]:
dados_limpos = preenche_tabela(dados)
a_remover = dados_limpos.query("WINDOW=='0-2' and ICU==1")['PATIENT_VISIT_IDENTIFIER'].values
dados_limpos = dados_limpos.query("PATIENT_VISIT_IDENTIFIER not in @a_remover")



In [35]:
dados_limpos

PATIENT_VISIT_IDENTIFIER  AGE_ABOVE65 AGE_PERCENTIL  GENDER  \
0                            0            1          60th       0   
1                            0            1          60th       0   
2                            0            1          60th       0   
3                            0            1          60th       0   
4                            0            1          60th       0   
...                        ...          ...           ...     ...   
1920                       384            0          50th       1   
1921                       384            0          50th       1   
1922                       384            0          50th       1   
1923                       384            0          50th       1   
1924                       384            0          50th       1   

      DISEASE GROUPING 1  DISEASE GROUPING 2  DISEASE GROUPING 3  \
0                    0.0                 0.0                 0.0   
1                    0.0                 0.0                 0.0   
2                    0.0                 0.0                 0.0   
3                    0.0                 0.0                 0.0   
4                    0.0                 0.0                 0.0   
...                  ...                 ...                 ...   
1920                 0.0                 0.0                 0.0   
1921                 0.0                 0.0                 0.0   
1922                 0.0                 0.0                 0.0   
1923                 0.0                 0.0                 0.0   
1924                 0.0                 0.0                 1.0   

      DISEASE GROUPING 4  DISEASE GROUPING 5  DISEASE GROUPING 6  ...  \
0                    0.0                 1.0                 1.0  ...   
1                    0.0                 1.0                 1.0  ...   
2                    0.0                 1.0                 1.0  ...   
3                    0.0                 1.0                 1.0  ...   
4                    0.0                 1.0                 1.0  ...   
...                  ...                 ...                 ...  ...   
1920                 0.0                 0.0                 0.0  ...   
1921                 0.0                 0.0                 0.0  ...   
1922                 0.0                 0.0                 0.0  ...   
1923                 0.0                 0.0                 0.0  ...   
1924                 0.0                 0.0                 0.0  ...   

      TEMPERATURE_DIFF  OXYGEN_SATURATION_DIFF  \
0            -1.000000               -1.000000   
1            -1.000000               -1.000000   
2            -1.000000               -1.000000   
3            -1.000000               -1.000000   
4            -0.238095               -0.818182   
...                ...                     ...   
1920         -1.000000               -1.000000   
1921         -1.000000               -1.000000   
1922         -1.000000               -1.000000   
1923         -1.000000               -1.000000   
1924         -0.547619               -0.838384   

      BLOODPRESSURE_DIASTOLIC_DIFF_REL  BLOODPRESSURE_SISTOLIC_DIFF_REL  \
0                            -1.000000                        -1.000000   
1                            -1.000000                        -1.000000   
2                            -1.000000                        -1.000000   
3                            -1.000000                        -1.000000   
4                            -0.389967                         0.407558   
...                                ...                              ...   
1920                         -1.000000                        -1.000000   
1921                         -1.000000                        -1.000000   
1922                         -1.000000                        -1.000000   
1923                         -1.000000                        -1.000000   
1924                         -0.701863                   

Agora serão selecionados os dados referentes apenas a janela 0-2, para que o modelo posteriormente implementado, consiga fazer uma previsão precoce sobre a necessidade ou não de UTI.Também vou excluir as linhas com dados nulos que ainda possam ter ficado no dataset.

In [36]:
def prepare_window(rows):
    if(np.any(rows["ICU"])):
        rows.loc[rows["WINDOW"]=="0-2", "ICU"] = 1
    return rows.loc[rows["WINDOW"] == "0-2"]

dados_limpos = dados_limpos.groupby("PATIENT_VISIT_IDENTIFIER").apply(prepare_window)

In [37]:
dados_limpos = dados_limpos.dropna()

In [38]:
dados_limpos

PATIENT_VISIT_IDENTIFIER  AGE_ABOVE65  \
PATIENT_VISIT_IDENTIFIER                                               
0                        0                            0            1   
2                        10                           2            0   
3                        15                           3            0   
4                        20                           4            0   
5                        25                           5            0   
...                                                 ...          ...   
380                      1900                       380            0   
381                      1905                       381            1   
382                      1910                       382            0   
383                      1915                       383            0   
384                      1920                       384            0   

                              AGE_PERCENTIL  GENDER  DISEASE GROUPING 1  \
PATIENT_VISIT_IDENTIFIER                                                  
0                        0             60th       0                 0.0   
2                        10            10th       0                 0.0   
3                        15            40th       1                 0.0   
4                        20            10th       0                 0.0   
5                        25            10th       0                 0.0   
...                                     ...     ...                 ...   
380                      1900          40th       1                 0.0   
381                      1905    Above 90th       0                 0.0   
382                      1910          50th       0                 0.0   
383                      1915          40th       1                 0.0   
384                      1920          50th       1                 0.0   

                               DISEASE GROUPING 2  DISEASE GROUPING 3  \
PATIENT_VISIT_IDENTIFIER                                                
0                        0                    0.0                 0.0   
2                        10                   0.0                 0.0   
3                        15                   0.0                 0.0   
4                        20                   0.0                 0.0   
5                        25                   0.0                 0.0   
...                                           ...                 ...   
380                      1900                 0.0                 0.0   
381                      1905                 0.0                 0.0   
382                      1910                 0.0                 0.0   
383                      1915                 0.0                 0.0   
384                      1920                 0.0                 0.0   

                               DISEASE GROUPING 4  DISEASE GROUPING 5  \
PATIENT_VISIT_IDENTIFIER                                                
0                        0                    0.0                 1.0   
2                        10                   0.0                 0.0   
3                        15                   0.0                 0.0   
4                        20                   0.0                 0.0   
5                        25                   0.0                 0.0   
...                                           ...                 ...   
380                      1900                 0.0                 0.0   
381                      1905                 0.0                 0.0   
382                      1910                 0.0                 0.0   
383                      1915                 0.0                 0.0   
384                      1920                 0.0                 0.0   

                               DISEASE GROUPING 6  ...  TEMPERATURE_DIFF  \
PATIENT_VISIT_IDENTIFIER                           ...                     
0                        0                    1.0  ...         -1.

Por fim, as colunas PATIENT_VISIT_IDENTIFIER  e WINDOW serão deletadas. Isso porque essas colunas não influenciam no paciente ir ou não para UTI. Além disso, estou utilizando os dados de apenas uma das janelas(0-2).

In [28]:
#Salvando os dados limpos em um arquivo csv na pasta de dados do projeto
dados_limpos.drop(['PATIENT_VISIT_IDENTIFIER', 'WINDOW'], axis=1).to_csv('/Users/camil/Jupyter Projects/dados/dados_limpos.csv', index= False)